test/ZNSStringResolver.test.ts

```diff
@@ -0,0 +1,464 @@
+import { SignerWithAddress } from "@nomicfoundation/hardhat-ethers/signers";
+import * as hre from "hardhat";
+import {
+  ADMIN_ROLE,
+  DEFAULT_PRICE_CONFIG,
+  DEFAULT_ROYALTY_FRACTION,
+  distrConfigEmpty,
+  getAccessRevertMsg,
+  GOVERNOR_ROLE,
+  hashDomainLabel,
+  INITIALIZED_ERR,
+  paymentConfigEmpty,
+  REGISTRAR_ROLE,
+  validateUpgrade,
+  ZNS_DOMAIN_TOKEN_NAME,
+  ZNS_DOMAIN_TOKEN_SYMBOL,
+} from "./helpers";
+import { MeowMainnet } from "../src/deploy/missions/contracts/meow-token/mainnet-data";
+import { IZNSContracts } from "../src/deploy/campaign/types";
+import { runZnsCampaign } from "../src/deploy/zns-campaign";
+import { expect } from "chai";
+import * as ethers from "ethers";
+import { registrationWithSetup } from "./helpers/register-setup";
+import {
+  ERC165__factory,
+  MeowTokenMock, ZNSAccessController, ZNSDomainToken, ZNSRegistry, ZNSRootRegistrar,
+  ZNSStringResolver,
+  ZNSStringResolverUpgradeMock__factory,
+  ZNSTreasury,
+} from "../typechain";
+import { DeployCampaign, MongoDBAdapter } from "@zero-tech/zdc";
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+import { DefenderRelayProvider } from "@openzeppelin/defender-sdk-relay-signer-client/lib/ethers";
+import { getConfig } from "../src/deploy/campaign/environments";
+
+
+describe("ZNSStringResolver", () => {
+  describe("One campaign for everybody", () => {
+    let zeroVault : SignerWithAddress;
+    let operator : SignerWithAddress;
+    let user : SignerWithAddress;
+    let deployAdmin : SignerWithAddress;
+    let deployer : SignerWithAddress;
+    let admin : SignerWithAddress;
+
+    const env = "dev";
+
+    let stringResolver : ZNSStringResolver;
+    let registry : ZNSRegistry;
+    let campaign : DeployCampaign<
+    HardhatRuntimeEnvironment,
+    SignerWithAddress,
+    DefenderRelayProvider,
+    IZNSContracts
+    >;
+    let rootRegistrar : ZNSRootRegistrar;
+
+    let userBalance : bigint;
+
+    const uri = "https://example.com/817c64af";
+    const domainName = "lox";
+    const domainNameHash = hashDomainLabel(domainName);
+
+    let mongoAdapter : MongoDBAdapter;
+
+    before(async () => {
+
+      [
+        deployer,
+        zeroVault,
+        user,
+        deployAdmin,
+        admin,
+        operator,
+      ] = await hre.ethers.getSigners();
+
+      const campaignConfig = await getConfig({
+        deployer: deployer as unknown as SignerWithAddress,
```

Comment:
this probably doesn't need cast to `unknown`. can just cast `as SignerWithAddress` right away, without `unknown`